### PR TITLE
Upgrade the IDF version from 4.4 to 5.0 (MAJOR CHANGE)

### PR DESCRIPTION
**Pull request by rnd-ash**

### CODE REVIEW

File: src/diag/perf_mon.cpp [hunk 1]
```diff
@@ -87,7 +82,7 @@ esp_err_t init_perfmon(void)
 
 void remove_perfmon(void)
 {
-    timer_deinit(TIMER_GROUP_0, TIMER_1);
+    //gptimer_disable(gptimer);
```
Comment (chrissivo) [inline src/diag/perf_mon.cpp:85]: I would expect that line not to be a comment.

File: src/sensors.cpp [hunk 2]
```diff
@@ -1,261 +1,201 @@
...
 #include "esp_log.h"
 #include "board_config.h"
 #include "nvs/eeprom_config.h"
 #include "esp_check.h"
+#include "driver/gptimer.h"
+#include "driver/pulse_cnt.h"
 
 #define PULSES_PER_REV 60 // N2 and N3 are 60 pulses per revolution
-#define PCNT_H_LIM 1
+#define MAX_RPM_PCNT 10000
```
Comment (chrissivo) [inline src/sensors.cpp:16]: I would suggest to make this a configurable value.

File: src/sensors.cpp [hunk 3]
```diff
@@ -1,261 +1,201 @@
...
-    time_delta_n3 = n3_intr_times[1] - n3_intr_times[0];
-    portEXIT_CRITICAL_ISR(&n3_mux);
-    portENTER_CRITICAL_ISR(&output_mux);
-    time_delta_output = output_intr_times[1] - output_intr_times[0];
-    portEXIT_CRITICAL_ISR(&output_mux);
+static bool IRAM_ATTR on_rpm_timer(gptimer_handle_t timer, const gptimer_alarm_event_data_t *edata, void *user_data) {
+    int pulses = 0;
+    // N2 Sensor
+    if (ESP_OK == read_and_reset_pcnt(PCNT_HANDLE_N2, &pulses)) {
+        n2_total = n2_total - n2_avgs[avg_n2_idx];
```
Comment (chrissivo) [inline src/sensors.cpp:74]: I would suggest to write "n2_total -= n2_avgs[avg_n2_idx];". It's in general safer von multi threading environments and avoids to use a wrong variable.

File: src/sensors.cpp [hunk 4]
```diff
@@ -1,261 +1,201 @@
...
 
-    if (time_delta_n3 != 0) {
-        n3_total -= n3_avgs[avg_n3_idx];
-        n3_avgs[avg_n3_idx] = time_delta_n3;
-        n3_total += time_delta_n3;
-        avg_n3_idx = (avg_n3_idx+1)%RPM_SAMPLES_DEBOUNCE;
     }
+    // N3 Sensor
+    if (ESP_OK == read_and_reset_pcnt(PCNT_HANDLE_N3, &pulses)) {
+        n3_total = n3_total - n3_avgs[avg_n3_idx];
```
Comment (chrissivo) [inline src/sensors.cpp:83]: The deleted equivalent already uses the proposal given above.

File: src/sensors.cpp [hunk 5]
```diff
@@ -1,261 +1,201 @@
...
-        output_total -= output_avgs[avg_out_idx];
-        output_avgs[avg_out_idx] = time_delta_output;
-        output_total += time_delta_output;
+    // Output Sensor (If present)
+    if (output_rpm_ok && ESP_OK == read_and_reset_pcnt(PCNT_HANDLE_OUTPUT, &pulses)) {
+        output_total = output_total - output_avgs[avg_out_idx];
+        output_avgs[avg_out_idx] = pulses;
+        output_total = output_total + pulses;
         avg_out_idx = (avg_out_idx+1)%RPM_SAMPLES_DEBOUNCE;
+    } else {
```
Comment (chrissivo) [inline src/sensors.cpp:96]: Since else doesn't contain commands, it could be deleted.

File: src/sensors.cpp [hunk 6]
```diff
@@ -1,261 +1,201 @@
...
+
     }
+    return true;
```
Comment (chrissivo) [inline src/sensors.cpp:99]: Since the function always returns true, it could also be of type void.

File: src/sensors.cpp [hunk 7]
```diff
@@ -1,261 +1,201 @@
...
+        "Failed to set PCNT actions for %s PCNT",
+        name
+    );
+    ESP_RETURN_ON_ERROR(pcnt_unit_set_glitch_filter(*UNIT_HANDLE, &glitch_filter), "SENSORS", "Failed to set glitch filter for PCNT unit %s", name);
+    ESP_RETURN_ON_ERROR(pcnt_unit_enable(*UNIT_HANDLE), "SENSORS", "Failed to enable PCNT unit %s", name);
+    ESP_RETURN_ON_ERROR(pcnt_unit_start(*UNIT_HANDLE), "SENSORS", "Failed to start PCNT unit %s", name);
+    return ESP_OK;
+}
 
+esp_err_t Sensors::init_sensors(){
```
Comment (chrissivo) [inline src/sensors.cpp:134]: from MISRA: init_sensors() should be init_sensors(void)

File: src/sensors.cpp [hunk 8]
```diff
@@ -1,261 +1,201 @@
...
-        .counter_h_lim = PCNT_H_LIM,
-        .counter_l_lim = 0,
-        .unit = PCNT_N3_RPM,
-        .channel = PCNT_CHANNEL_0};
-
+const pcnt_glitch_filter_config_t glitch_filter = {
+    .max_glitch_ns = 1000
+};
+
+esp_err_t configure_pcnt(const char* name, gpio_num_t gpio, pcnt_unit_handle_t* UNIT_HANDLE, pcnt_channel_handle_t* CHANNEL_HANDLE) {
```
Comment (chrissivo) [inline src/sensors.cpp:106]: from MISRA: A function should have a single point of exit.

File: src/torque_converter.cpp [hunk 9]
```diff
@@ -85,10 +85,10 @@ void TorqueConverter::update(GearboxGear curr_gear, PressureManager* pm, Abstrac
             this->was_idle = false;
             if (this->tcc_learn_lockup_map != nullptr) {
                 this->curr_tcc_target = this->tcc_learn_lockup_map->get_value((float)curr_gear, 1.0);
-                ESP_LOGI("TCC", "Learn cell value is %d mBar", curr_tcc_target);
+                ESP_LOGI("TCC", "Learn cell value is %lu mBar", curr_tcc_target);
                 this->initial_ramp_done = false;
-                this->base_tcc_pressure = MAX(0, this->curr_tcc_target-500);
-                this->curr_tcc_pressure = MAX(0, this->curr_tcc_target-500);
+                this->base_tcc_pressure = MAX(0, this->curr_tcc_target-300);
```
Comment (chrissivo) [inline src/torque_converter.cpp:90]: I would rather suggest to define these values ("300") as central variables/parameters instead of magic numbers.

**Review by chrissivo — APPROVED** [review]
Have found only minor things, which are partially related to a preferred programming style. However, the changes in general look good and reasonable to me.

File: src/diag/perf_mon.cpp (hunk 1 above)
Comment (rnd-ash) [reply src/diag/perf_mon.cpp]: Your right! Resolved in d3ad19d22fae508a6f375b3a12e583e511b7925a

File: src/sensors.cpp (hunk 3 above)
Comment (rnd-ash) [reply src/sensors.cpp:74]: actually no. with n2_total (And similar) being volatile (Due to syncing between cores depending on which core handles the interrupt), -= assignment is deprecated in C++20 with volatile. So this is the only way to do it without making the compiler complain.

File: src/torque_converter.cpp (hunk 9 above)
Comment (rnd-ash) [reply src/torque_converter.cpp:90]: Yes true, I will work on that back in the dev branch once the merge is completed.

File: src/sensors.cpp (hunk 6 above)
Comment (rnd-ash) [reply src/sensors.cpp:99]: See the docs [here](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/gptimer.html), the return true is needed in this ISR context.

File: src/sensors.cpp (hunk 2 above)
Comment (rnd-ash) [reply src/sensors.cpp:16]: True. I think we should discuss at a later date (In Dev branch), having a settings configuration for variables in the code. Not necessarily a map data structure, but almost something like "Advanced options" for debugging purposes only.

File: src/sensors.cpp (hunk 5 above)
Comment (rnd-ash) [reply src/sensors.cpp]: Fixed in https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/gptimer.html

File: src/sensors.cpp (hunk 7 above)
Comment (rnd-ash) [reply src/sensors.cpp]: Fixed in https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/gptimer.html

**Pull request merged**